File: USR-CoreFab/src/orchestration.rs
//! Constitutional Orchestration — Lifecycle and communication between components.
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use crate::{Result, ConstitutionalError, timestamp};

/// Component lifecycle state
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum LifecycleState {
    Registered,
    Initialized,
    Running,
    Paused,
    Stopped,
    Error,
}

/// A message between components
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub from: String,
    pub to: String,
    pub payload: String,
    pub timestamp: String,
}

/// Component state tracking
#[derive(Debug, Clone)]
struct ComponentState {
    state: LifecycleState,
    registered_at: String,
    last_transition: String,
}

/// The orchestration engine — manages component lifecycle and communication
pub struct OrchestrationEngine {
    components: HashMap<String, ComponentState>,
    messages: Vec<Message>,
}

impl OrchestrationEngine {
    pub fn new() -> Self {
        Self {
            components: HashMap::new(),
            messages: Vec::new(),
        }
    }

    /// Register a component with the runtime
    pub fn register(&mut self, component_id: &str) -> Result<()> {
        if self.components.contains_key(component_id) {
            return Err(ConstitutionalError::OrchestrationError(
                format!("Component {} already registered", component_id)
            ));
        }

        self.components.insert(component_id.to_string(), ComponentState {
            state: LifecycleState::Registered,
            registered_at: timestamp(),
            last_transition: timestamp(),
        });

        Ok(())
    }

    /// Initialize a registered component
    pub fn initialize(&mut self, component_id: &str) -> Result<()> {
        let state = self.components.get_mut(component_id)
            .ok_or_else(|| ConstitutionalError::OrchestrationError(
                format!("Component {} not found", component_id)
            ))?;

        if state.state != LifecycleState::Registered {
            return Err(ConstitutionalError::OrchestrationError(
                format!("Component {} cannot initialize from state {:?}", component_id, state.state)
            ));
        }

        state.state = LifecycleState::Initialized;
        state.last_transition = timestamp();
        Ok(())
    }

    /// Start an initialized component
    pub fn start(&mut self, component_id: &str) -> Result<()> {
        let state = self.components.get_mut(component_id)
            .ok_or_else(|| ConstitutionalError::OrchestrationError(
                format!("Component {} not found", component_id)
            ))?;

        if state.state != LifecycleState::Initialized {
            return Err(ConstitutionalError::OrchestrationError(
                format!("Component {} cannot start from state {:?}", component_id, state.state)
            ));
        }

        state.state = LifecycleState::Running;
        state.last_transition = timestamp();
        Ok(())
    }

    /// Pause a running component
    pub fn pause(&mut self, component_id: &str) -> Result<()> {
        let state = self.components.get_mut(component_id)
            .ok_or_else(|| ConstitutionalError::OrchestrationError(
                format!("Component {} not found", component_id)
            ))?;

        if state.state != LifecycleState::Running {
            return Err(ConstitutionalError::OrchestrationError(
                format!("Component {} cannot pause from state {:?}", component_id, state.state)
            ));
        }

        state.state = LifecycleState::Paused;
        state.last_transition = timestamp();
        Ok(())
    }

    /// Stop a component
    pub fn stop(&mut self, component_id: &str) -> Result<()> {
        let state = self.components.get_mut(component_id)
            .ok_or_else(|| ConstitutionalError::OrchestrationError(
                format!("Component {} not found", component_id)
            ))?;

        if state.state == LifecycleState::Stopped {
            return Err(ConstitutionalError::OrchestrationError(
                format!("Component {} already stopped", component_id)
            ));
        }

        state.state = LifecycleState::Stopped;
        state.last_transition = timestamp();
        Ok(())
    }

    /// Get the current state of a component
    pub fn get_state(&self, component_id: &str) -> Result<LifecycleState> {
        self.components.get(component_id)
            .map(|s| s.state.clone())
            .ok_or_else(|| ConstitutionalError::OrchestrationError(
                format!("Component {} not found", component_id)
            ))
    }

    /// Send a message between components
    pub fn send_message(&mut self, from: &str, to: &str, payload: &str) -> Result<()> {
        // Verify both components exist and are running
        let from_state = self.get_state(from)?;
        let to_state = self.get_state(to)?;

        if from_state != LifecycleState::Running {
            return Err(ConstitutionalError::OrchestrationError(
                format!("Sender {} is not running (state: {:?})", from, from_state)
            ));
        }

        if to_state != LifecycleState::Running {
            return Err(ConstitutionalError::OrchestrationError(
                format!("Recipient {} is not running (state: {:?})", to, to_state)
            ));
        }

        self.messages.push(Message {
            from: from.to_string(),
            to: to.to_string(),
            payload: payload.to_string(),
            timestamp: timestamp(),
        });

        Ok(())
    }

    /// Get all messages for a component
    pub fn get_messages(&self, component_id: &str) -> Vec<&Message> {
        self.messages
            .iter()
            .filter(|m| m.to == component_id || m.from == component_id)
            .collect()
    }

    /// Get the dependency graph as JSON
    pub fn get_dependency_graph(&self) -> String {
        let graph: Vec<serde_json::Value> = self.components.iter().map(|(id, state)| {
            serde_json::json!({
                "id": id,
                "state": format!("{:?}", state.state),
                "registered_at": state.registered_at,
                "last_transition": state.last_transition,
            })
        }).collect();

        serde_json::to_string_pretty(&graph).unwrap_or_default()
    }

    /// Verify all components are in valid states
    pub fn verify_all(&self) -> Result<bool> {
        for (id, state) in &self.components {
            match state.state {
                LifecycleState::Registered | LifecycleState::Initialized |
                LifecycleState::Running | LifecycleState::Paused | LifecycleState::Stopped => {
                    // Valid states
                }
                LifecycleState::Error => {
                    return Err(ConstitutionalError::OrchestrationError(
                        format!("Component {} is in error state", id)
                    ));
                }
            }
        }
        Ok(true)
    }
}

File: usr/engine/orchestration.js
// USR/CoreFab — Orchestration Contract (C6)
// Constitutional Orchestration Engine — lifecycle and communication.

const LifecycleState = {
  Registered: 'registered',
  Initialized: 'initialized',
  Running: 'running',
  Paused: 'paused',
  Stopped: 'stopped',
  Error: 'error',
};

const VALID_TRANSITIONS = {
  [LifecycleState.Registered]: [LifecycleState.Initialized, LifecycleState.Error],
  [LifecycleState.Initialized]: [LifecycleState.Running, LifecycleState.Error],
  [LifecycleState.Running]: [LifecycleState.Paused, LifecycleState.Stopped, LifecycleState.Error],
  [LifecycleState.Paused]: [LifecycleState.Running, LifecycleState.Stopped, LifecycleState.Error],
  [LifecycleState.Stopped]: [],
  [LifecycleState.Error]: [],
};

class OrchestrationEngine {
  constructor() {
    this.components = new Map();
    this.messages = [];
  }

  register(componentId) {
    if (this.components.has(componentId)) {
      throw new Error(`OrchestrationError: Component ${componentId} already registered`);
    }
    this.components.set(componentId, {
      state: LifecycleState.Registered,
      registeredAt: timestamp(),
      lastTransition: timestamp(),
    });
    window.dispatchEvent(new CustomEvent('usr:orchestration:registered', { detail: { componentId } }));
  }

  _transition(componentId, targetState) {
    const comp = this.components.get(componentId);
    if (!comp) throw new Error(`OrchestrationError: Component ${componentId} not found`);
    const valid = VALID_TRANSITIONS[comp.state] || [];
    if (!valid.includes(targetState)) {
      throw new Error(`OrchestrationError: Cannot transition ${componentId} from ${comp.state} to ${targetState}`);
    }
    comp.state = targetState;
    comp.lastTransition = timestamp();
    window.dispatchEvent(new CustomEvent('usr:orchestration:transition', { detail: { componentId, from: comp.state, to: targetState } }));
  }

  initialize(componentId) { this._transition(componentId, LifecycleState.Initialized); }
  start(componentId) { this._transition(componentId, LifecycleState.Running); }
  pause(componentId) { this._transition(componentId, LifecycleState.Paused); }
  stop(componentId) { this._transition(componentId, LifecycleState.Stopped); }

  getState(componentId) {
    const comp = this.components.get(componentId);
    if (!comp) throw new Error(`OrchestrationError: Component ${componentId} not found`);
    return comp.state;
  }

  sendMessage(from, to, payload) {
    const fromComp = this.components.get(from);
    const toComp = this.components.get(to);
    if (!fromComp) throw new Error(`OrchestrationError: Sender ${from} not found`);
    if (!toComp) throw new Error(`OrchestrationError: Recipient ${to} not found`);
    if (fromComp.state !== LifecycleState.Running) throw new Error(`OrchestrationError: Sender ${from} not running`);
    if (toComp.state !== LifecycleState.Running) throw new Error(`OrchestrationError: Recipient ${to} not running`);

    const msg = { from, to, payload, timestamp: timestamp() };
    this.messages.push(msg);
    window.dispatchEvent(new CustomEvent('usr:orchestration:message', { detail: { message: msg } }));
    return msg;
  }

  getMessages(componentId) {
    return this.messages.filter(m => m.to === componentId || m.from === componentId);
  }

  getDependencyGraph() {
    return Array.from(this.components.entries()).map(([id, state]) => ({
      id, state: state.state, registeredAt: state.registeredAt, lastTransition: state.lastTransition,
    }));
  }

  async verifyAll() {
    for (const [id, comp] of this.components) {
      if (comp.state === LifecycleState.Error) {
        throw new Error(`OrchestrationError: Component ${id} is in error state`);
      }
    }
    return true;
  }
}

window.OrchestrationEngine = OrchestrationEngine;
window.LifecycleState = LifecycleState;
